🔒 Added a "reset all passwords" feature

tpatel opened the pull request:
issue https://github.com/TryGhost/Team/issues/750

- Only accessible by admins
- Resets all staff users' passwords and prevents them to log-in
- Sends them a reset email password to give them back access to their account
- Closes all existing staff user sessions

tpatel committed:
🔒 Added a "reset all passwords" feature

issue https://github.com/TryGhost/Team/issues/750

- Only accessible by admins
- Resets all staff users' passwords and prevents them to log-in
- Sends them a reset email password to give them back access to their account
- Closes all existing staff user sessions

naz commented on core/server/web/api/canary/admin/routes.js:
```diff
@@ -67,6 +67,7 @@ module.exports = function apiRoutes() {
 
     // ## Users
     router.get('/users', mw.authAdminApi, http(apiCanary.users.browse));
+    router.get('/users/reset-all-passwords', mw.authAdminApi, http(apiCanary.users.resetAllPasswords));
```
I don't think we use kebab casing anywhere else to name an endpoint :thinking:  It's usually underscore separated like: `/email_preview/posts/:id`,  `/members/stats/gross_volume`.

Also not sure about the name. What other names have you considered when picking this one?

matthanley replied:
This should also **definitely not** be a GET route

tpatel replied:
It's updated to `PUT /users/reset_all_passwords`

tpatel replied:
Regarding the name, this is how we're talking about the feature in the designs so it feels like the most descriptive name, I'm open to suggestions though.

![image](https://user-images.githubusercontent.com/494686/120668278-9a6a7400-c48e-11eb-94a3-165a40bb4ce5.png)

naz replied:
Imo "all" part is not necessary for the non-user facing naming. Naming it `reset_password` leaves a door open to do things like passing a list of emails that have been compromised to this endpoint and doing a "bulk reset password". Also, worth giving a thought if we want to thing of a "password" as a separate resource that happens to belong to a user or leave it as is :thinking:

naz replied:
@tpatel to add to above comment I'm not sure the `PUT` is a correct method here either. I struggled to find a good reference. The best example we use often is Stipe's API. In [similar "action" situation](https://stripe.com/docs/api/setup_intents) they use `POST`:
```
  POST /v1/setup_intents/:id/confirm
  POST /v1/setup_intents/:id/cancel
```

What outside internal references did you check when deciding on the naming?

tpatel replied:
I had started by looking at the internal usage of `POST`/`PUT` in Ghost to mimic the choices made by the team:

```
PUT /users/password => changePassword
PUT /authentication/passwordreset => resetPassword
POST /authentication/passwordreset => generateResetToken
```

As we're modifying the state of users, sending a password reset email AND deleting sessions, it's hard to know which one fits best and I arbitrarily chose `PUT`.

Now, reading [RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9) seems to point towards using `POST` instead. The main reason is that `PUT` is mentioned as being idempotent vs `POST` not having to be idempotent.

I've updated it ✅

naz replied:
Yup. Makes sense :+1:

naz commented on core/server/api/canary/users.js:
```diff
@@ -230,5 +233,31 @@ module.exports = {
+
+    resetAllPasswords: {
+        headers: {
+            cacheInvalidate: true
```
Why are we invalidating the cache?

tpatel replied:
I couldn't find any docs on this property, I was thinking it would avoid the specific route from landing in the cache.
After looking through the Ghost Pro repositories, it looks like this header might clear the cache for the whole Ghost instance?

naz replied:
> this header might clear the cache for the whole Ghost instance?

Yup, that's exactly what it does. We don't want to do that unless any of the content (posts, tags, etc.) has changed.

naz commented on core/server/api/canary/users.js:
```diff
@@ -230,5 +233,31 @@ module.exports = {
+        async query(frame) {
+            await userService.resetAllPasswords(frame.options);
+
+            // TODO: make sure it's ok to return this kind of JSON
```
Would be good to have references for this from other endpoints that are "actions" and not resource fetching.

tpatel replied:
The only other instances I could find were [these](https://github.com/TryGhost/Ghost/blob/f4017cc4392afee117a29ad11d47a76e3883799e/core/server/api/canary/members.js#L346
) [two](https://github.com/TryGhost/Ghost/blob/273e22032761d39aef10e0b9a90584f31b0271c9/core/server/api/canary/webhooks.js#L146) so I'm changing this to `null` .

naz replied:
Yup. This is way better! :+1:

naz commented on core/server/api/canary/users.js:
```diff
@@ -230,5 +233,31 @@ module.exports = {
+        },
+        permissions: async (frame) => {
+            // TODO: Is there an easier way than querying the roles?
+            const user = await models.User.findOne({
```
Is there a related resource which has similar permission level needed? We might need to add entries to permissions table instead :thinking:  Definitely should not be doing checks like this here and leave it for the permission layer to handle  :wink:

naz commented on core/server/services/users.js:
```diff
@@ -17,17 +19,66 @@ const path = require('path');
+            const users = await this.models.User.findAll({...frameOptions});
+            for (const user of users) {
+                await user.save({
+                    password: await Users.randomPassword(),
```
Do we even need to set a random password or just setting the locked status is enough?

tpatel replied:
good catch, the locked status should be enough!

naz reviewed (commented):
Added quite few comments. I think the direction is good needs some more reference material and polishing up (it's not `main` ready for sure :sweat_smile: )

matthanley commented on core/server/web/api/canary/admin/routes.js:
```diff
@@ -75,6 +75,7 @@ module.exports = function apiRoutes() {
 
     router.put('/users/password', mw.authAdminApi, http(apiCanary.users.changePassword));
     router.put('/users/owner', mw.authAdminApi, http(apiCanary.users.transferOwnership));
+    router.put('/users/reset_all_passwords', mw.authAdminApi, http(apiCanary.users.resetAllPasswords));
```
@tpatel imo this would make more sense as an auth endpoint like `/authentication/reset` - we already have `passwordreset` there: https://github.com/TryGhost/Ghost/blob/5ef9bdc91a36dbf46a413ce70b2f34e6e6fc495f/core/server/web/api/canary/admin/routes.js#L211

tpatel replied:
Yes I was hesitating between the two with no strong opinion, I'm gonna move it under `/authentication` then 👍

naz commented on core/server/services/users.js:
```diff
@@ -17,17 +17,59 @@ const path = require('path');
+            frameOptions.transacting = t;
+
+            // Reset all passwords
+            const users = await this.models.User.findAll({...frameOptions});
```
What's the difference of using `...` spread operator to just doing a `.findAll(frameOptions)` call?

tpatel replied:
I've fixed it, I guess I was expecting to add more properties to the findAll configuration object but didn't end up doing so.

naz commented on core/server/services/users.js:
```diff
@@ -17,17 +17,59 @@ const path = require('path');
+                }, {...frameOptions});
+            }
+
+            // Delete all sessions
```
I don't think this whole block belongs here. Anything to do with session data processing imo should be in the [SessionStore](https://github.com/TryGhost/Ghost/blob/44053553f5baf7a2a1e2f688ceef235afb9e3b82/core/server/services/auth/session/store.js). 

The session store base class we extend in Ghost has exactly the method that is mean to remove all the session data - https://github.com/expressjs/session/tree/0048bcac451ad867299d404aca94c79cc8bc751d#storeclearcallback.

Sidenote, maybe worth looking into truncating the table instead of doing a findAll-> iterate delete dance :thinking:

tpatel replied:
The two reasons I had chosen the findAll->iterate are:
1. It's only touching the staff users, so it shouldn't be more than a couple hundreds sessions
2. This way it triggers hooks if we want to use them in the future.

I'm fine with doing a truncate though, it's optimizing performance over the hooks feature which is ok.

Regarding the code placement, I've moved everything to the express session object but I had to export the express out of the session service, so let me know what you think about it.

naz commented on core/server/data/schema/fixtures/fixtures.json:
```diff
@@ -486,6 +486,11 @@
+                },
+                {
+                    "name": "Reset all passwords",
+                    "action_type": "resetAllPasswords",
```
I'd give the permission naming a hard think here. What alternative names have been considered here?

tpatel replied:
Here is the full list of what's used as an `action_type` value: `"activate", "add", "assign", "auth", "backupContent", "browse", "deleteAllContent", "destroy", "download", "edit", "exportContent", "generate", "importContent", "publish", "read", "retry", "send", "sendTestEmail", "upload"`. With this context I feel it's ok to name the action that resets all passwords `resetAllPasswords` 😅 
Do you have an example of what would be more fitting?

naz replied:
Yup! Now it totally makes sense, specially that we have `deleteAllContent` :+1:

tpatel committed:
Moved the session reset to the express session store

pr https://github.com/TryGhost/Ghost/pull/13005

naz commented on core/server/models/session.js:
```diff
@@ -48,6 +48,12 @@ const Session = ghostBookshelf.Model.extend({
 
+    destroyAll() {
+        // As of 2021, it's the recommended way to truncate a table
+        // https://github.com/bookshelf/bookshelf/issues/1740
```
Noice :ok:

naz commented on core/server/services/auth/session/express-session.js:
```diff
@@ -22,7 +23,8 @@ const expressSessionMiddleware = session({
 });
 
-exports.getSession = async function getSession(req, res) {
+module.exports = {};
```
Is this line needed? :thinking:

tpatel replied:
nope, TIL 😄

naz commented on core/server/services/auth/session/index.js:
```diff
@@ -49,3 +49,4 @@ module.exports.createSessionFromToken = sessionFromToken({
 });
 
 module.exports.sessionService = sessionService;
+module.exports.sessionStore = expressSession.sessionStore;
```
I was thinking about this bit and I think a cleaner approach here would be exposing a separate method on the `sessionsService` like `clear()` that would call the `store.clear()`. "Store" is an internal implementation detail of this module, don't think controllers should know about it. The other upside is having way smaller exposed API surface with a method vs whole Store instance. Wdyt?

tpatel replied:
I didn't think about this approach, I'm looking into it now 👀

tpatel replied:
After a couple of attempts I think it's cleaner not to bring the concept of express-session into our already-extracted session library.
So I've simply reduced the surface area of what's exposed by only exposing a promisified version of `store.clear()`, I think it's the best of both worlds

naz replied:
Love it!

tpatel committed:
Updated the reset_all_passwords method to use the http post method

pr https://github.com/TryGhost/Ghost/pull/13005

tpatel committed:
Removed sessionStore from being exposed

pr https://github.com/TryGhost/Ghost/pull/13005

Also fixed the resetAllPasswords so that it actually waits after we deleted all sessions.

naz commented on core/server/api/canary/authentication.js:
```diff
@@ -190,5 +197,14 @@ module.exports = {
+    },
+
+    resetAllPasswords: {
+        // validation: false,
```
@tpatel what's the reason behind this being commented out?

tpatel replied:
It's definitely a leftover from early attempts at learning the Ghost framework, I've removed this because uncommenting would skip the validation step and there is no data to validate!

naz commented on core/server/data/migrations/versions/4.7/03-add-reset-all-passwords-permission.js:
```diff
@@ -0,0 +1,11 @@
+const {
```
@tpatel this file needs a bump as the next version is 4.8 I think ;)

naz reviewed (approved):
@tpatel I think it's good to go just few minor tweaks needed: 
1. Change the migration folder (and maybe the number?)
2. Make sure to write a test for this endpoint! This can be done post merge, but we have to have at least basic test coverage :wink:

tpatel committed:
Added a regression test for the reset-all-passwords feature

pr https://github.com/TryGhost/Ghost/pull/13005

tpatel commented:
I've updated the migration (most likely I'll have to update it again before merging) and I've added one regression test!

I'm thinking it's not a good thing to have it in the regression test folder though, I might have to convert that to a set of unit tests instead.
